src/v/cluster/plugin_frontend.cc

```diff
@@ -0,0 +1,403 @@
+        case raft::errc::timeout:
+            return errc::timeout;
+        case raft::errc::not_leader:
+            return errc::not_leader_controller;
+        default:
+            return errc::replication_error;
+        }
+    }
+    if (ec.category() == rpc::error_category()) {
+        switch (rpc::errc(ec.value())) {
+        case rpc::errc::client_request_timeout:
+            return errc::timeout;
+        default:
+            return errc::replication_error;
+        }
+    }
+    if (ec.category() == error_category()) {
+        return errc(ec.value());
+    }
+    return errc::replication_error;
+}
+} // namespace
+
+plugin_frontend::plugin_frontend(
+  model::node_id s,
+  partition_leaders_table* l,
+  plugin_table* t,
+  topic_table* tp,
+  controller_stm* c,
+  rpc::connection_cache* r,
+  ss::abort_source* a)
+  : _self(s)
+  , _leaders(l)
+  , _connections(r)
+  , _table(t)
+  , _topics(tp)
+  , _abort_source(a)
+  , _controller(c) {}
+
+ss::future<errc> plugin_frontend::upsert_transform(
+  transform_metadata meta, model::timeout_clock::time_point timeout) {
+    // The ID is looked up later.
+    transform_cmd c{transform_update_cmd{transform_id(-1), std::move(meta)}};
+    return do_mutation(std::move(c), timeout).then([](auto r) { return r.ec; });
+}
+
+ss::future<mutation_result> plugin_frontend::remove_transform(
+  transform_name name, model::timeout_clock::time_point timeout) {
+    transform_cmd c{transform_remove_cmd{std::move(name), 0}};
+    return do_mutation(std::move(c), timeout);
+}
+
+ss::future<mutation_result> plugin_frontend::do_mutation(
+  transform_cmd cmd, model::timeout_clock::time_point timeout) {
+    auto cluster_leader = _leaders->get_leader(model::controller_ntp);
+    if (!cluster_leader) {
+        co_return mutation_result{.ec = errc::no_leader_controller};
+    }
+    if (*cluster_leader != _self) {
+        co_return co_await dispatch_mutation_to_remote(
+          *cluster_leader,
+          std::move(cmd),
+          timeout - model::timeout_clock::now());
+    }
+    if (ss::this_shard_id() != controller_stm_shard) {
+        co_return co_await container().invoke_on(
+          controller_stm_shard,
+          // NOLINTNEXTLINE(cppcoreguidelines-avoid-reference-coroutine-parameters)
+          [cmd = std::move(cmd), timeout](auto& service) mutable {
+              return service.do_mutation(std::move(cmd), timeout);
+          });
+    }
+    // Make sure we're up to date
+    auto result = co_await _controller->quorum_write_empty_batch(timeout);
+    if (!result) {
+        co_return mutation_result{.ec = errc::not_leader_controller};
+    }
+    mutation_result r = co_await do_local_mutation(
+      std::move(cmd), result.value().last_offset, timeout);
+    if (r.ec != errc::success) {
+        co_return r;
+    }
+    // This is an optimization to reduce metadata propagation lag, we can
+    // safely ignore the result as it doesn't effect correctness.
```

Comment:
not sure I followed this, what metadata propagation is this referring to?

Reply:
Sorry I am copying this from https://github.com/redpanda-data/redpanda/blob/7fa9368d77ebc4144f4382b47384c89af2798de3/src/v/cluster/topics_frontend.cc#L201-L207

Should I remove?

Reply:
I think it can be removed, also the barrier _after_ do_local_mutation() can also be removed I think, not sure what purpose it is serving.. @mmaslankaprv has more context on the topics_frontend code, Michal, wdyt.

Reply:
I removed it, we talked offline this is to speed metadata propagation to clients by the time they respond, so clients are less likely to get inconsistent views. It's unlikely we need this for wasm deploys, so I've removed it.